### PR TITLE
operator: add gantry as a cluster component (prototype)

### PR DESCRIPTION
**Pull request by plombardi89**
## Summary

Prototype that registers **gantry** (peer-to-peer OCI distribution agent) as a cluster component in the unbounded-operator, built on the component-registry interface from #516. It demonstrates that a new `ClusterComponent` plugs in with a single package + one registry line, and that the operator can deploy gantry end-to-end (workload + node containerd wiring).

> Stacked on `operator-component-registry` (#516). Review that first; this PR's base is that branch, so the diff here is only the gantry addition. It rebases onto main once #516 merges.

## What it does

Gantry runs one agent DaemonSet per node fronting containerd image pulls, so it maps to a cluster singleton (like net/machina). It is registered in `DefaultRegistry().Cluster` and reconciled by the generic driver - no driver/status/watch plumbing changes.

The operator installs both halves:
- **Workload**: the gantry agent DaemonSet + ServiceAccount/RBAC + PriorityClass + `gantry-config` ConfigMap.
- **Node wiring**: the `gantry-containerd-config` DaemonSet + `gantry-containerd-hosts` ConfigMap that write `/etc/containerd/certs.d/_default/hosts.toml`, routing node pulls through the local mirror.

## Changes

- **`deploy/gantry`**: add `embed.go` (`//go:embed all:rendered`) so the manifests are importable; templatize the DaemonSet image (was a hardcoded `ghcr.io/vpatelsj/gantry:latest`) so the render stamps a version-matched tag; rename `node-config.yaml` -> `node-config.yaml.tmpl` and template its namespace (was a hardcoded `gantry-system`) so it renders/embeds and follows the install namespace.
- **Makefile**: pass `--set Image=$(GANTRY_IMAGE)` to `gantry-manifests` and add it to the `test`/`build`/`vulncheck`/`unbounded-operator-build` render prerequisites (+ empty-image guard), so the operator embeds real, non-`:latest` gantry manifests.
- **Operator RBAC**: grant `scheduling.k8s.io/priorityclasses` so the operator can install the `gantry-low` PriorityClass (+ an `rbac_test.go` assertion).
- **API**: add `spec.components.gantry`. Unlike the other components, **gantry defaults to enabled** (opt-out): its `EnabledFor` treats an absent/unset spec as enabled. `make generate` updated deepcopy + the Site CRD.
- **`internal/operator/components/gantry`**: `ClusterComponent` + `WatchProvider` mirroring `net` - config seed-or-preserve, config-hash stamping on the agent and node-config DaemonSets, watches on the config ConfigMaps + both DaemonSets. Applies only the top-level manifests; the `examples/` subtree (optional NetworkPolicy, sample Secret) is excluded from the default install via a new `Env.ApplyManifestFiles` primitive.

## Caveats (documented, not fixed here)

- Node wiring is only effective where containerd's CRI `config_path=/etc/containerd/certs.d` - set by the agent (`pkg/agent/phases/nodestart/assets/containerd.toml`) on unbounded-provisioned nodes; a harmless no-op elsewhere.
- `certs.d/_default` routes every registry through the local mirror first; gantry accelerates configured `upstream_registries` and passes others back to origin fallback.
- `unbounded-system` must permit privileged/hostPath pods (already required today by net-node `hostNetwork` and storage hostPath); no component sets PSA labels, so this is a pre-existing namespace-wide assumption, not new to gantry.
- Retention: disabling gantry leaves the workload and node-config in place (containerd then dials a dead `127.0.0.1:5000`, ~200ms fallback per pull) until an explicit uninstall flow - a separate follow-up.

## Testing

- New gantry package tests: default-enabled `EnabledFor`, config seed/preserve, config-hash stamping (agent + node-config DaemonSets), retained-when-opted-out, and that the default install applies the core objects (ServiceAccount/DaemonSet/PriorityClass/RBAC + node-config ConfigMap/DaemonSet) but **not** the `examples/` NetworkPolicy/Secret or the separately-reconciled `gantry-config`.
- `manifests_guard_test` now covers gantry (`:latest` guard).
- `make gantry-manifests`, `make fmt`, `make lint` (0 issues), `go build ./...`, `go test ./internal/operator/... ./api/machina/...` all pass.

### CODE REVIEW

**Review by cchildress — DISMISSED** [review]

File: internal/operator/components/gantry/gantry.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+// Package gantry implements the gantry peer-to-peer OCI distribution cluster
+// component. Gantry runs one agent DaemonSet per node fronting containerd's
+// image pulls, so it is reconciled as a cluster singleton. Unlike the other
+// components it defaults to enabled: it is deployed unless every Site explicitly
+// opts out via spec.components.gantry.enabled=false.
+//
+// Note: deploying the gantry workload does not by itself route node containerd
+// through the mirror; that requires a per-node containerd hosts.toml under
+// /etc/containerd/certs.d, which is not yet managed by the operator.
```
[the remark this answers is not in the document]
Comment (vpatelsj) [reply internal/operator/components/gantry/gantry.go]: Fixed in 460e5423. The package comment now describes the managed node-config DaemonSet and the hosts.toml wiring it installs.

File: internal/operator/components/gantry/gantry.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+}
+
+func resourcesExist(ctx context.Context, env *component.Env) (bool, error) {
+	for _, resource := range []struct {
+		name   string
+		object client.Object
+	}{
+		{name: configName, object: &corev1.ConfigMap{}},
+		{name: daemonSetName, object: &appsv1.DaemonSet{}},
+	} {
```
[the remark this answers is not in the document]
Comment (vpatelsj) [reply internal/operator/components/gantry/gantry.go:170]: Fixed in 460e5423. Retention detection now includes gantry-containerd-hosts and gantry-containerd-config, with regression tests proving each partial-install resource triggers full reconciliation.

File: internal/operator/components/gantry/gantry.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+		// singleton removal is surprising. A future explicit uninstall flow
+		// should handle removal.
+		retained, err := resourcesExist(ctx, env)
+		if err != nil {
+			return component.Failed(err)
+		}
+
+		if !retained {
+			return component.Disabled("no site enables gantry; retained")
+		}
```
[the remark this answers is not in the document]
Comment (vpatelsj) [reply internal/operator/components/gantry/gantry.go]: Fixed in 460e5423. The disabled condition now says "no site enables gantry", and the result message is covered by a regression test.

File: deploy/unbounded-operator/02-rbac.yaml.tmpl [hunk 4]
```diff
@@ -51,6 +51,11 @@ rules:
   - apiGroups: ["batch"]
     resources: ["jobs", "cronjobs"]
     verbs: ["get", "list", "watch"]
+  - apiGroups: ["scheduling.k8s.io"]
+    # The gantry component installs a low-priority PriorityClass (gantry-low)
+    # so its agents are preemptible under cluster pressure.
+    resources: ["priorityclasses"]
+    verbs: ["get", "list", "watch", "create", "patch", "update", "delete", "deletecollection"]
```
[the remark this answers is not in the document]
Comment (vpatelsj) [reply deploy/unbounded-operator/02-rbac.yaml.tmpl]: Fixed in 460e5423. Removed delete and deletecollection from the PriorityClass rule; the RBAC test now also asserts those privileges remain absent.

File: internal/operator/component/env.go [hunk 5]
```diff
@@ -104,6 +104,46 @@ func SetPodSpecImages(obj *unstructured.Unstructured, image string) error {
...
+
+		if changed {
+			if err := unstructured.SetNestedSlice(obj.Object, containers, "spec", "template", "spec", field); err != nil {
+				return fmt.Errorf("set %s: %w", field, err)
+			}
+		}
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (vpatelsj) [reply internal/operator/component/env.go:152]: Fixed in 460e5423. SetNamedContainerImage now returns an error when the named container is absent; tests cover both targeted replacement and fail-closed behavior.

**Review by vpatelsj — APPROVED** [review]

File: Makefile [hunk 6]
```diff
@@ -652,7 +652,8 @@ gantry-manifests: ## Render gantry deployment manifests into deploy/gantry/rende
 	$(GOCMD) run ./hack/cmd/render-manifests \
 		--templates-dir $(GANTRY_MANIFEST_TEMPLATES_DIR) \
 		--output-dir $(GANTRY_MANIFEST_RENDERED_DIR) \
-		--set Namespace=$(GANTRY_NAMESPACE)
+		--set Namespace=$(GANTRY_NAMESPACE) \
+		--set Image=$(GANTRY_IMAGE)
```
[the remark this answers is not in the document]
Comment (vpatelsj) [reply Makefile:656]: The normal release path supplies a valid image, and the operator replaces the embedded image during reconciliation.

**Review by jwilder — APPROVED** [review]
